docs/GuidGuidRecommender.md

```diff
@@ -0,0 +1,144 @@
+for add-ons which are similar to a given add-on,
+where similarity is based on the number of profiles
+which have both add-ons installed.
+
+As recommendations are only surfaced
+for a specific set of add-ons meeting certain criteria,
+and depend only on characteristics of the add-ons themselves,
+recommendation sets for each add-on can be (and are) computed in advance.
+
+Recommendations are generated as follows:
+
+1. The [input data](#input-data),
+    consisting of coinstallation counts for each pair of add-ons,
+    is loaded from a precomputed file.
+2. For each add-on, we determine a [__candidate list__](#generating-the-scored-candidate-lists) of "related" add-ons,
+    together with associated __relevance scores__,
+    based on the coinstallation data.
+3. Recommendations for a given add-on are [selected](#selecting-recommendations)
+    by retaining the N add-ons in the candidate list
+    with the highest relevance scores.
+
+Once the recommender has been trained,
+we apply various [metrics](./GraphRepresentation.md#quality-and-health-metrics)
+to the coinstallation data, candidate lists, and recommendations,
+to evaluate the overall health of the system
+and the quality of recommendations.
+
+The recommender system is implemented by the [`GuidGuidCoinstallRecommender`](../taar_lite/recommenders/guidguid.py#L10) class,
+which is instantiated by the [TaarLiteAppResource](../taar_lite/app/production.py#L44).
+
+Further technical details are provided on [this page](./GraphRepresentation.md).
+
+
+# Input data
+
+The primary data source used in generating recommendations
+is the file `guid_coinstallation.json`,
+which is generated by the ETL jobs described in the [README](../README.md).
+It contains coinstallation counts for each pair of add-ons
+belonging to a filtered list.
+
+Coinstallation counts are computed based on the Telemetry data
+submitted by a random sample of recently active Firefox profiles.
+For each pair of add-ons in the list,
+we count the number of profiles in which both add-ons are installed and enabled.
+
+The data is stored in "sparse matrix representation" as a nested object/dict,
+in which each add-on GUID maps to a further object
+mapping coinstalled GUIDS to their coinstallation count,
+dropping zero counts.
+On initializing a [`GuidGuidCoinstallRecommender`](../taar_lite/recommenders/guidguid.py#L10),
+it is supplied as `raw_coinstall_dict`.
+
+Using this dataset as the basis for recommendations
+entails the following assumptions:
+
+- The set of add-ons appearing in the coinstallation dataset represent
+    the universe of known add-ons for the purpose of generating recommendations.
+- We are only dealing with add-ons that were coinstalled
+    with at least one other add-on on the list
+    (otherwise they would not appear in the file at all).
+- Profiles with only a single add-on installed do not contribute to the counts.
+
+__TODO:__ How are system add-ons handled? Add a note about this.
```

Comment:
Open an issue for this TODO

Reply:
Ah, they are specifically excluded: https://github.com/mozilla/python_mozetl/blob/4bd8ceed909ae779826c383cbf7659b5dbf5869b/mozetl/taar/taar_lite_guidguid.py#L63

In fact, I don't think system add-ons would be listed on AMO either.

Reply:
I'll add a comment about this in the README section on that script instead.